README.md
# Showcase
![pimp_my_button-showcase](media/pimp_my_gif.gif "pimp_my_button-showcase")

# Features

- Highly customizable (Don't like my particle effects? Make your own with little effort!)
- Very easy to use
- A lot of premade particles

# Installation

I will publish this package on pub if people enjoy it. So leave a star if you do ;)

```yaml
dependencies:
  pimp_my_button: ^1.0.0
```
run packages get and import it
```
import 'package:pimp_my_button/pimp_my_button.dart';
```
**If you are not able to import it, restart your IDE see: https://github.com/flutter/flutter/issues/17016**

# Basics

The basic setup looks like this:

```dart
PimpedButton(
  particle: DemoParticle(),
  pimpedWidgetBuilder: (context, controller) {
    return FloatingActionButton(onPressed: () {
      controller.forward(from: 0.0);
    },);
  },
);
```
The `pimpedWidgetBuilder` uses a builder which besides providing a new context and an `AnimationController`.
When your button is pressed call this code:

```dart
controller.forward(from: 0.0);
```
It's important to include the `from` field because otherwise the animation won't play after the first tap.

The particle specified in the `PimpedButton` controlls what the animation looks like.

# Demo Particle Walk Through

We'll walk through how to build a `Particle` yourself, step by step. 

Here is what were are going to build:

![fab](media/fab.gif "fab")


Each particle has to extend `Particle` and override this method:
`void paint(Canvas canvas, Size size, progress, seed)`
- Canvas is the canvas you can paint on
- The size is the bounding box of the the enclosing widget
- Progress is a double between 0 and 1, it reflects the progress in the animation
- The seed is an int value generated once for every tap. When dealing with randoms, initialize your
`Random` with that seed (so every frame is synced)

## The first particle
```dart
CompositeParticle(
  children: []
).paint(canvas, size, progressm seed);
```
You have to paint atleast one `Particle` in the paint(.,.,.,) method. Usually this would be the `CompositeParticle`.

The `CompositeParticle` doesn't do much on its own. It only paints all of its children.


## Adding actual particles

The next interesting particle is the `CircleMirror` and `RectangeMirror`.

```dart
    CircleMirror(
        numberOfParticles: 6,
        child: AnimatedPositionedParticle(
          begin: Offset(0.0, 20.0),
          end: Offset(0.0, 60.0),
          child: FadingRect(width: 5.0, height: 15.0, color: Colors.pink),
        ),
        initialRotation: -pi / randomMirrorOffset),
```
This mirrors its particle around the middle point in a circular shape.
In this case you provide one `Particle` which is going to be drawn multiple times, thus looking
identical. If you want different particles (or the same with randomized values), use the `CircleMirror.builder`.

At the bottom of the hierarchy is the `FadingRect`, all it does is drawing a rectangle which fades out over time. 
To make it move, it's wrapped in an `AnimatedPositionedParticle`.

This is the full code:
```dart
class DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      CircleMirror(
          numberOfParticles: 6,
          child: AnimatedPositionedParticle(
            begin: Offset(0.0, 20.0),
            end: Offset(0.0, 60.0),
            child: FadingRect(width: 5.0, height: 15.0, color: Colors.pink),
          ),
          initialRotation: -pi / randomMirrorOffset),
      CircleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (index) {
            return IntervalParticle(
                child: AnimatedPositionedParticle(
                  begin: Offset(0.0, 30.0),
                  end: Offset(0.0, 50.0),
                  child: FadingTriangle(
                      baseSize: 6.0 + random.nextDouble() * 4.0,
                      heightToBaseFactor: 1.0 + random.nextDouble(),
                      variation: random.nextDouble(),
                      color: Colors.green),
                ),
                interval: Interval(
                  0.0,
                  0.8,
                ));
          },
          // division by 0 is not good ;)
          initialRotation: -pi / randomMirrorOffset + 8),
    ]).paint(canvas, size, progress, seed);
  }
}
```


# Notes

### Some particles worth checking out:
- `IntervalParticle`, applies an interval and/or curve to the child.
- `FourRandomSlotParticle`, positions 4 children on the different sections.
- `PoppingCircle`, pretty popping circle

### Check out other demo particles:
https://github.com/Norbert515/pimp_my_button/blob/master/lib/src/demo_particles.dart

### You are not restricted to use the particles in the context of "Pimp my button", the particles can be drawn on any regular canvas.





CHANGELOG.md
## [1.0.0] - Release on pub

Pimp my button on pub


example/lib/main.dart

import 'package:pimp_my_button/pimp_my_button.dart';
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text(widget.title),
      ),
      body: Center(
        child: Column(
          children: <Widget>[
            Expanded(
              child: Row(
                children: <Widget>[
                  Expanded(
                    child: Center(
                      child: PimpedButton(
                        particle: DemoParticle(),
                        pimpedWidgetBuilder: (context, controller) {
                          return FloatingActionButton(onPressed: () {
                            controller.forward(from: 0.0);
                          },);
                        },
                      ),
                    ),
                  ),
                  Expanded(
                    child: Center(
                      child: PimpedButton(
                        particle: RectangleDemoParticle(),
                        pimpedWidgetBuilder: (context, controller) {
                          return RaisedButton(onPressed: () {
                            controller.forward(from: 0.0);
                          },
                            child: Text("Special button"),
                          );
                        },
                      ),
                    ),
                  ),
                  Expanded(
                    child: Center(
                      child: PimpedButton(
                        particle: Rectangle2DemoParticle(),
                        pimpedWidgetBuilder: (context, controller) {
                          return MaterialButton(onPressed: () {
                            controller.forward(from: 0.0);
                          },
                            child: Text("Special button"),
                          );
                        },
                      ),
                    ),
                  ),
                ],
              ),
            ),
            PimpedButton(
              particle: ListTileDemoParticle(),
              pimpedWidgetBuilder: (context, controller) {
                return ListTile(
                  title: Text("ListTile"),
                  subtitle: Text("Some nice subtitle"),
                  trailing: Icon(Icons.add),
                  onTap: () {
                    controller.forward(from: 0.0);
                  },
                );
              },
            ),
            Center(
              child: PimpedButton(
                particle: Rectangle2DemoParticle(),
                pimpedWidgetBuilder: (context, controller) {
                  return IconButton(
                    icon: Icon(Icons.favorite_border),
                    color: Colors.indigo,
                    onPressed: () {
                      controller.forward(from: 0.0);
                    },
                  );
                },
              ),
            ),
            Expanded(
              child: Center(
                child: PimpedButton(
                  particle: Rectangle3DemoParticle(),
                  pimpedWidgetBuilder: (context, controller) {
                    return RaisedButton(onPressed: () {
                      controller.forward(from: 0.0);
                    },
                      child: Text("Rectangles"),
                    );
                  },
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}



lib/src/demo_particles.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:pimp_my_button/src/pimp_my_button.dart';

Color intToColor(int col) {
  col = col % 5;
  if (col == 0) return Colors.red;
  if (col == 1) return Colors.green;
  if (col == 2) return Colors.orange;
  if (col == 3) return Colors.blue;
  if (col == 4) return Colors.pink;
  if (col == 5) return Colors.brown;
  return Colors.black;
}

class DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      CircleMirror(
          numberOfParticles: 6,
          child: AnimatedPositionedParticle(
            begin: Offset(0.0, 20.0),
            end: Offset(0.0, 60.0),
            child: FadingRect(width: 5.0, height: 15.0, color: Colors.pink),
          ),
          initialRotation: -pi / randomMirrorOffset),
      CircleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (index) {
            return IntervalParticle(
                child: AnimatedPositionedParticle(
                  begin: Offset(0.0, 30.0),
                  end: Offset(0.0, 50.0),
                  child: FadingTriangle(
                      baseSize: 6.0 + random.nextDouble() * 4.0,
                      heightToBaseFactor: 1.0 + random.nextDouble(),
                      variation: random.nextDouble(),
                      color: Colors.green),
                ),
                interval: Interval(
                  0.0,
                  0.8,
                ));
          },
          initialRotation: -pi / randomMirrorOffset + 8),
    ]).paint(canvas, size, progress, seed);
  }
}

class RectangleDemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 13,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -10.0),
              end: Offset(0.0, -60.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: -pi / randomMirrorOffset),
      CircleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (index) {
            return IntervalParticle(
                child: AnimatedPositionedParticle(
                  begin: Offset(0.0, 30.0),
                  end: Offset(0.0, 50.0 + (7.5 - 15 * random.nextDouble())),
                  child: FadingTriangle(
                      baseSize: 6.0 + random.nextDouble() * 4.0,
                      heightToBaseFactor: 1.0 + random.nextDouble(),
                      variation: random.nextDouble(),
                      color: Colors.green),
                ),
                interval: Interval(
                  0.0,
                  0.8,
                ));
          },
          initialRotation: -pi / randomMirrorOffset + 8),
    ]).paint(canvas, size, progress, seed);
  }
}

class Rectangle2DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -10.0),
              end: Offset(0.0, -60.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: -pi / randomMirrorOffset),
    ]).paint(canvas, size, progress, seed);
  }
}

class Rectangle3DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
                begin: Offset(0.0, -10.0),
                end: Offset(0.0, -50.0),
                child: RotationParticle(
                  rotation: random.nextDouble() * (2 * pi),
                  child: FadingTriangle(
                    baseSize: 12.0 + random.nextDouble(),
                    heightToBaseFactor: 0.8 + random.nextDouble(),
                    variation: random.nextDouble(),
                    color: intToColor(int),
                  ),
                ));
          },
          initialDistance: -pi / randomMirrorOffset),
      RectangleMirror.builder(
          numberOfParticles: 8,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
                begin: Offset(0.0, -10.0),
                end: Offset(0.0, -30.0),
                child: RotationParticle(
                  rotation: random.nextDouble() * (2 * pi),
                  child: FadingTriangle(
                    baseSize: 12.0 + random.nextDouble(),
                    heightToBaseFactor: 0.8 + random.nextDouble(),
                    variation: random.nextDouble(),
                    color: intToColor(int),
                  ),
                ));
          },
          initialDistance: 80.0),
    ]).paint(canvas, size, progress, seed);
  }

  double randomOffset(Random random, int range) {
    return range / 2 - random.nextInt(range);
  }
}

class ListTileDemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    CompositeParticle(children: [
      Firework(),
      Firework(),
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 8,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -30.0),
              end: Offset(0.0, -80.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: 0.0),
      RectangleMirror.builder(
          numberOfParticles: 5,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -25.0),
              end: Offset(0.0, -60.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: 30.0),
      RectangleMirror.builder(
          numberOfParticles: 8,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -40.0),
              end: Offset(0.0, -100.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: 80.0),
    ]).paint(canvas, size, progress, seed);
  }

}


example/lib/particles.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:pimp_my_button/pimp_my_button.dart';

Color intToColor(int col) {
  col = col % 5;
  if (col == 0) return Colors.red;
  if (col == 1) return Colors.green;
  if (col == 2) return Colors.orange;
  if (col == 3) return Colors.blue;
  if (col == 4) return Colors.pink;
  if (col == 5) return Colors.brown;
  return Colors.black;
}

class DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      CircleMirror(
          numberOfParticles: 6,
          child: AnimatedPositionedParticle(
            begin: Offset(0.0, 20.0),
            end: Offset(0.0, 60.0),
            child: FadingRect(width: 5.0, height: 15.0, color: Colors.pink),
          ),
          initialRotation: -pi / randomMirrorOffset),
      CircleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (index) {
            return IntervalParticle(
                child: AnimatedPositionedParticle(
                  begin: Offset(0.0, 30.0),
                  end: Offset(0.0, 50.0),
                  child: FadingTriangle(
                      baseSize: 6.0 + random.nextDouble() * 4.0,
                      heightToBaseFactor: 1.0 + random.nextDouble(),
                      variation: random.nextDouble(),
                      color: Colors.green),
                ),
                interval: Interval(
                  0.0,
                  0.8,
                ));
          },
          initialRotation: -pi / randomMirrorOffset + 8),
    ]).paint(canvas, size, progress, seed);
  }
}

class RectangleDemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 13,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -10.0),
              end: Offset(0.0, -60.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: -pi / randomMirrorOffset),
      CircleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (index) {
            return IntervalParticle(
                child: AnimatedPositionedParticle(
                  begin: Offset(0.0, 30.0),
                  end: Offset(0.0, 50.0 + (7.5 - 15 * random.nextDouble())),
                  child: FadingTriangle(
                      baseSize: 6.0 + random.nextDouble() * 4.0,
                      heightToBaseFactor: 1.0 + random.nextDouble(),
                      variation: random.nextDouble(),
                      color: Colors.green),
                ),
                interval: Interval(
                  0.0,
                  0.8,
                ));
          },
          initialRotation: -pi / randomMirrorOffset + 8),
    ]).paint(canvas, size, progress, seed);
  }
}

class Rectangle2DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -10.0),
              end: Offset(0.0, -60.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: -pi / randomMirrorOffset),
    ]).paint(canvas, size, progress, seed);
  }
}

class Rectangle3DemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    int randomMirrorOffset = random.nextInt(8) + 1;
    CompositeParticle(children: [
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 6,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
                begin: Offset(0.0, -10.0),
                end: Offset(0.0, -50.0),
                child: RotationParticle(
                  rotation: random.nextDouble() * (2 * pi),
                  child: FadingTriangle(
                    baseSize: 12.0 + random.nextDouble(),
                    heightToBaseFactor: 0.8 + random.nextDouble(),
                    variation: random.nextDouble(),
                    color: intToColor(int),
                  ),
                ));
          },
          initialDistance: -pi / randomMirrorOffset),
      RectangleMirror.builder(
          numberOfParticles: 8,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
                begin: Offset(0.0, -10.0),
                end: Offset(0.0, -30.0),
                child: RotationParticle(
                  rotation: random.nextDouble() * (2 * pi),
                  child: FadingTriangle(
                    baseSize: 12.0 + random.nextDouble(),
                    heightToBaseFactor: 0.8 + random.nextDouble(),
                    variation: random.nextDouble(),
                    color: intToColor(int),
                  ),
                ));
          },
          initialDistance: 80.0),
    ]).paint(canvas, size, progress, seed);
  }

  double randomOffset(Random random, int range) {
    return range / 2 - random.nextInt(range);
  }
}

class ListTileDemoParticle extends Particle {
  @override
  void paint(Canvas canvas, Size size, progress, seed) {
    Random random = Random(seed);
    CompositeParticle(children: [
      Firework(),
      Firework(),
      Firework(),
      RectangleMirror.builder(
          numberOfParticles: 8,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -30.0),
              end: Offset(0.0, -80.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: 0.0),
      RectangleMirror.builder(
          numberOfParticles: 5,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -25.0),
              end: Offset(0.0, -60.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: 30.0),
      RectangleMirror.builder(
          numberOfParticles: 8,
          particleBuilder: (int) {
            return AnimatedPositionedParticle(
              begin: Offset(0.0, -40.0),
              end: Offset(0.0, -100.0),
              child: FadingRect(width: 5.0, height: 15.0, color: intToColor(int)),
            );
          },
          initialDistance: 80.0),
    ]).paint(canvas, size, progress, seed);
  }

}
